pallets/subbridge/src/dynamic_trader.rs

```diff
@@ -0,0 +1,116 @@
+use assets_registry::GetAssetRegistryInfo;
+use frame_support::{
+	traits::Get,
+	weights::{constants::WEIGHT_PER_SECOND, Weight},
+};
+use sp_std::{marker::PhantomData, result::Result, vec::Vec};
+use xcm::latest::{prelude::*, AssetId};
+use xcm_builder::TakeRevenue;
+use xcm_executor::{traits::WeightTrader, Assets};
+
+pub struct DynamicWeightTrader<
+	WeightPerSecond,
+	FungibleAssetId,
+	FungibleAssetsInfo: GetAssetRegistryInfo<FungibleAssetId>,
+	R: TakeRevenue,
+>(
+	Weight,
+	u128,
+	Option<(AssetId, u128)>,
+	PhantomData<(WeightPerSecond, FungibleAssetId, FungibleAssetsInfo, R)>,
+);
+impl<WeightPerSecond, FungibleAssetId, FungibleAssetsInfo, R> WeightTrader
+	for DynamicWeightTrader<WeightPerSecond, FungibleAssetId, FungibleAssetsInfo, R>
+where
+	WeightPerSecond: Get<u64>,
+	FungibleAssetsInfo: GetAssetRegistryInfo<FungibleAssetId>,
+	R: TakeRevenue,
+{
+	fn new() -> Self {
+		Self(0, 0, None, PhantomData)
+	}
+
+	fn buy_weight(&mut self, weight: Weight, payment: Assets) -> Result<Assets, XcmError> {
```

Comment:
Please double check if saturating operations are good. Maybe better to use checked operations?

Reply:
The above line ```let unused = payment.checked_sub((id, amount).into()).map_err(|_| XcmError::TooExpensive)?;``` already has overflow check, following `saturating_xxx` can work properly

Reply:
In such case, I think you can just use plain operator (`+`, `-`)

Reply:
Just follow the upstream [impl](https://github.com/paritytech/polkadot/blob/2c285771342a84e13dab696ef86902f73d4fbd4c/xcm/xcm-builder/src/weight.rs#L158)